File: backend/src/repositories/dynamodb-session.repository.ts
import {
  GetCommand,
  PutCommand,
  QueryCommand,
  type DynamoDBDocumentClient,
} from '@aws-sdk/lib-dynamodb';

import { BackendEnv } from '../config/env';
import { LoopSession } from '../models/session';
import {
  buildSessionLookupKeys,
  buildSessionPrimaryKey,
  buildSessionsByProjectKey,
  DYNAMODB_ENTITY_TYPES,
} from './dynamodb-key-design';
import { SessionRepository } from './session.repository';

type DynamoDbSessionItem = LoopSession & {
  pk: string;
  sk: string;
  entityType: typeof DYNAMODB_ENTITY_TYPES.session;
  sessionId: string;
  gsi2pk: string;
  gsi2sk: string;
};

type DynamoDbSessionRepositoryEnv = Pick<BackendEnv, 'DYNAMODB_METADATA_TABLE_NAME'>;

export class DynamoDbSessionRepository implements SessionRepository {
  constructor(
    private readonly client: DynamoDBDocumentClient,
    private readonly env: DynamoDbSessionRepositoryEnv
  ) {}

  async listSessions() {
    const response = await this.client.send(
      new QueryCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        IndexName: 'gsi2',
        KeyConditionExpression: 'gsi2sk = :gsi2sk',
        ExpressionAttributeValues: {
          ':gsi2sk': 'METADATA',
        },
      })
    );

    return (response.Items ?? []).map((item) => this.toSession(item as DynamoDbSessionItem));
  }

  async listSessionsByProject(projectId: string) {
    const keys = buildSessionsByProjectKey(projectId);

    const response = await this.client.send(
      new QueryCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        KeyConditionExpression: 'pk = :pk AND begins_with(sk, :skPrefix)',
        ExpressionAttributeValues: {
          ':pk': keys.pk,
          ':skPrefix': keys.skPrefix,
        },
      })
    );

    return (response.Items ?? []).map((item) => this.toSession(item as DynamoDbSessionItem));
  }

  async getSessionById(sessionId: string) {
    const lookupKeys = buildSessionLookupKeys(sessionId);

    const response = await this.client.send(
      new QueryCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        IndexName: 'gsi2',
        KeyConditionExpression: 'gsi2pk = :gsi2pk AND gsi2sk = :gsi2sk',
        ExpressionAttributeValues: {
          ':gsi2pk': lookupKeys.gsi2pk,
          ':gsi2sk': lookupKeys.gsi2sk,
        },
        Limit: 1,
      })
    );

    const item = response.Items?.[0];

    if (!item) {
      return null;
    }

    return this.toSession(item as DynamoDbSessionItem);
  }

  async createSession(session: LoopSession) {
    await this.client.send(
      new PutCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        Item: this.toItem(session),
        ConditionExpression: 'attribute_not_exists(pk) AND attribute_not_exists(sk)',
      })
    );

    return session;
  }

  async reset() {
    throw new Error('DynamoDB session repository reset is not supported.');
  }

  private toItem(session: LoopSession): DynamoDbSessionItem {
    return {
      ...buildSessionPrimaryKey(session.projectId, session.id),
      ...buildSessionLookupKeys(session.id),
      ...session,
      entityType: DYNAMODB_ENTITY_TYPES.session,
      sessionId: session.id,
    };
  }

  private toSession(item: DynamoDbSessionItem): LoopSession {
    return {
      id: item.sessionId,
      projectId: item.projectId,
      name: item.name,
      bpm: item.bpm,
      trackCount: item.trackCount,
      createdAt: item.createdAt,
      updatedAt: item.updatedAt,
    };
  }
}

File: backend-dotnet/src/Loopr.Api/Repositories/DynamoDbSessionRepository.cs
using Amazon.DynamoDBv2.Model;
using Loopr.Api.Configuration;
using Loopr.Api.Domain;
using Microsoft.Extensions.Options;

namespace Loopr.Api.Repositories;

public sealed class DynamoDbSessionRepository(
    IDynamoDbMetadataStore store,
    IOptions<DynamoDbOptions> options
) : ISessionRepository
{
    private const string LookupIndexName = "gsi2";

    private readonly string tableName = options.Value.MetadataTableName;

    public async Task<IReadOnlyList<LoopSession>> ListAsync(
        CancellationToken cancellationToken
    )
    {
        var response = await store.ScanAsync(
            new ScanRequest
            {
                TableName = tableName,
                FilterExpression = "entityType = :entityType",
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>
                {
                    [":entityType"] = DynamoDbAttributeMap.String(DynamoDbEntityTypes.Session),
                },
            },
            cancellationToken
        );

        return (response.Items ?? [])
            .Select(ToSession)
            .OrderByDescending(session => session.UpdatedAt)
            .ToArray();
    }

    public async Task<LoopSession?> GetByIdAsync(
        string sessionId,
        CancellationToken cancellationToken
    )
    {
        var lookupKey = DynamoDbMetadataKeys.BuildSessionLookupKey(sessionId);
        var response = await store.QueryAsync(
            new QueryRequest
            {
                TableName = tableName,
                IndexName = LookupIndexName,
                KeyConditionExpression = "gsi2pk = :gsi2pk AND gsi2sk = :gsi2sk",
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>
                {
                    [":gsi2pk"] = DynamoDbAttributeMap.String(lookupKey.PartitionKey),
                    [":gsi2sk"] = DynamoDbAttributeMap.String(lookupKey.SortKey),
                },
                Limit = 1,
            },
            cancellationToken
        );

        var item = (response.Items ?? []).FirstOrDefault();

        return item is null ? null : ToSession(item);
    }

    public async Task<LoopSession> CreateAsync(
        LoopSession session,
        CancellationToken cancellationToken
    )
    {
        await store.PutItemAsync(
            new PutItemRequest
            {
                TableName = tableName,
                Item = ToItem(session),
                ConditionExpression = "attribute_not_exists(pk) AND attribute_not_exists(sk)",
            },
            cancellationToken
        );

        return session;
    }

    public Task ResetAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        throw new NotSupportedException("DynamoDB session repository reset is not supported.");
    }

    private static Dictionary<string, AttributeValue> ToItem(LoopSession session)
    {
        var primaryKey = DynamoDbMetadataKeys.BuildSessionPrimaryKey(
            session.ProjectId,
            session.Id
        );
        var lookupKey = DynamoDbMetadataKeys.BuildSessionLookupKey(session.Id);

        return new Dictionary<string, AttributeValue>
        {
            ["pk"] = DynamoDbAttributeMap.String(primaryKey.PartitionKey),
            ["sk"] = DynamoDbAttributeMap.String(primaryKey.SortKey),
            ["entityType"] = DynamoDbAttributeMap.String(DynamoDbEntityTypes.Session),
            ["projectId"] = DynamoDbAttributeMap.String(session.ProjectId),
            ["sessionId"] = DynamoDbAttributeMap.String(session.Id),
            ["name"] = DynamoDbAttributeMap.String(session.Name),
            ["bpm"] = DynamoDbAttributeMap.Number(session.Bpm),
            ["trackCount"] = DynamoDbAttributeMap.Number(session.TrackCount),
            ["createdAt"] = DynamoDbAttributeMap.Timestamp(session.CreatedAt),
            ["updatedAt"] = DynamoDbAttributeMap.Timestamp(session.UpdatedAt),
            ["gsi2pk"] = DynamoDbAttributeMap.String(lookupKey.PartitionKey),
            ["gsi2sk"] = DynamoDbAttributeMap.String(lookupKey.SortKey),
        };
    }

    private static LoopSession ToSession(Dictionary<string, AttributeValue> item)
    {
        return new LoopSession(
            DynamoDbAttributeMap.ReadString(item, "sessionId"),
            DynamoDbAttributeMap.ReadString(item, "projectId"),
            DynamoDbAttributeMap.ReadString(item, "name"),
            DynamoDbAttributeMap.ReadInt(item, "bpm"),
            DynamoDbAttributeMap.ReadInt(item, "trackCount"),
            DynamoDbAttributeMap.ReadTimestamp(item, "createdAt"),
            DynamoDbAttributeMap.ReadTimestamp(item, "updatedAt")
        );
    }
}
